### PR TITLE
Synchronous refer params AIS-8

### PR DESCRIPTION
**Pull request by E-B-Smith**
I added the method `- (NSDictionary*) getLatestReferringParamsSynchronous` to the Branch shared object.

I created a companion branch, `synchronous-refer-params-AIS-8-test` that has extra testing code included in Branch-TestBed.

That branch should not be merged, but it's useful for testing this code.  That branch already includes this code.

### CODE REVIEW

**Comment by aaustin** [comment]
Hmm - can you pull in the latest changes from staging into this branch? It seems out of date with my fixed from last week, making it hard to review

**Comment by aaustin** [comment]
@ahmednawar When you have a spare cycle, can you help me take a look over this one too? Just worried I'll miss something.

File: Branch-SDK/Branch-SDK/BNCServerInterface.m [hunk 1]
```diff
@@ -138,10 +138,9 @@ - (void)genericHTTPRequest:(NSURLRequest *)request retryNumber:(NSInteger)retryN
             }
             
         }
-        dispatch_async(dispatch_get_main_queue(), ^{
-            if (callback)
-                callback(serverResponse, error);
-        });
+		//	Don't call on the main queue since it might be blocked.
+		if (callback)
```
Comment (aaustin) [inline Branch-SDK/Branch-SDK/BNCServerInterface.m]: weird tabbing on these lines

File: Branch-SDK/Branch-SDK/Requests/BranchOpenRequest.m [hunk 2]
```diff
@@ -173,4 +176,52 @@ - (NSString *)getActionName {
...
+            openRequestWaitQueueIsSuspended = YES;
+            dispatch_suspend(openRequestWaitQueue);
+        }
+    }
+}
+
++ (void) waitForOpenResponseLock {
+    NSLog(@"Wait for openRequestWaitQueue.");
+    dispatch_sync(openRequestWaitQueue, ^ {
+        NSLog(@"Finished waitForOpenResponseLock");
```
Comment (aaustin) [inline Branch-SDK/Branch-SDK/Requests/BranchOpenRequest.m]: can you clean up these logs?

Comment (aaustin) [inline Branch-SDK/Branch-SDK/Requests/BranchOpenRequest.m]: something feels gross about the syntax of this empty block. Maybe we could put a return in front of dispatch_sync and a return in the block for readability.

File: Branch-SDK/Branch-SDK/Requests/BranchOpenRequest.m [hunk 3]
```diff
@@ -173,4 +176,52 @@ - (NSString *)getActionName {
...
+    });
+}
+
++ (void) releaseOpenResponseLock {
+    @synchronized (self) {
```
Comment (aaustin) [inline Branch-SDK/Branch-SDK/Requests/BranchOpenRequest.m]: should we synchronize on the static queue or bool rather than the instance of `BranchOpenRequest`? Hypothetically, it seems like it's possible for multiple branch open objects to enter this block of code and potentially clash on the boolean, although I'm having trouble imagining the edge case that could cause this.

**Comment by aaustin** [comment]
Can you add a test case for when this is called in a non-main thread as well being called from the main thread?

File: Branch-SDK/Branch-SDK/Branch.m [hunk 4]
```diff
@@ -1181,16 +1197,17 @@ - (void)registerViewWithParams:(NSDictionary *)params andCallback:(callbackWithP
...
     }
 }
 
 - (void)applicationWillResignActive {
-    [self clearTimer];
-    self.sessionTimer = [NSTimer scheduledTimerWithTimeInterval:0.5 target:self selector:@selector(callClose) userInfo:nil repeats:NO];
+    self.appIsInBackground = YES;
+    [self callClose];
     [self.requestQueue persistImmediately];
+    [BranchOpenRequest setWaitNeededForOpenResponseLock];
```
Comment (aaustin) [inline Branch-SDK/Branch-SDK/Branch.m]: I don't understand why we need to call this here.

File: Branch-SDK/Branch-SDK/Branch.m [hunk 5]
```diff
@@ -1336,38 +1347,38 @@ - (void)initializeSession {
...
 
-    if ([self.requestQueue removeInstallOrOpen])
-        self.networkCount = 0;
-    BranchOpenRequest *req = [[clazz alloc] initWithCallback:initSessionCallback];
-    [self insertRequestAtFront:req];
-    [self processNextQueueItem];
+	@synchronized (self) {
+		if ([self.requestQueue removeInstallOrOpen])
+			self.networkCount = 0;
+		[BranchOpenRequest setWaitNeededForOpenResponseLock];
```
Comment (aaustin) [inline Branch-SDK/Branch-SDK/Branch.m]: I also don't understand why we need to call this here if it's called up in initialize?

File: Branch-SDK/Branch-SDK/Branch.m (hunk 4 above)
Comment (aaustin) [inline Branch-SDK/Branch-SDK/Branch.m]: Oh I think I see: because you increment it every time the open call completes.

File: Branch-SDK/Branch-SDK/Branch.m (hunk 5 above)
Comment (aaustin) [inline Branch-SDK/Branch-SDK/Branch.m]: Now I understand I think. If you call this here though, seems like you can get rid of the call in initialize.

File: Branch-SDK/Branch-SDK/Requests/BranchOpenRequest.m [hunk 6]
```diff
@@ -82,6 +82,7 @@ - (void)makeRequest:(BNCServerInterface *)serverInterface key:(NSString *)key ca
 
 - (void)processResponse:(BNCServerResponse *)response error:(NSError *)error {
     if (error) {
+        [BranchOpenRequest releaseOpenResponseLock];
```
Comment (aaustin) [inline Branch-SDK/Branch-SDK/Requests/BranchOpenRequest.m]: Does this cover the retry case? Where a request fails & returns a non-nil error, then triggers a retry call? Will the queue be re-incremented?

File: Branch-SDK/Branch-SDK/Requests/BranchOpenRequest.m (hunk 3 above)
Comment (aaustin) [inline Branch-SDK/Branch-SDK/Requests/BranchOpenRequest.m]: Ah maybe because this is called on the static instance, it's fine.

File: Branch-SDK/Branch-SDK/Branch.m [hunk 7]
```diff
@@ -94,6 +93,7 @@ @interface Branch() <BranchDeepLinkingControllerCompletionDelegate, FABKit>
 @property (assign, nonatomic) BOOL delayForAppleAds;
 @property (assign, nonatomic) BOOL searchAdsDebugMode;
 @property (strong, nonatomic) NSMutableArray *whiteListedSchemeList;
+@property (assign, nonatomic) BOOL appIsInBackground;
```
Comment (aaustin) [inline Branch-SDK/Branch-SDK/Branch.m]: I think this boolean is unused? right?

File: Branch-SDK/Branch-SDK/Requests/BranchOpenRequest.m (hunk 6 above)
Comment (E-B-Smith) [inline Branch-SDK/Branch-SDK/Requests/BranchOpenRequest.m]: By the time `processResponse:error:` is called, the network call has exceeded the retries.

I want to unlock the lock so that the client app won't freeze.  

But if there's a network error, Branch isn't really initialized correctly, and `getLatestReferringParamsSynchronous` will return stale information.  Maybe the `getLatestReferringParamsSynchronous` should return a dictionary like:

```
{
    error = BNCErrorNetworkError
}
```

to indicate that there was a problem?

File: Branch-SDK/Branch-SDK/Branch.m (hunk 7 above)
Comment (E-B-Smith) [inline Branch-SDK/Branch-SDK/Branch.m]: True.  Better synchronization of the initialization made this unnecessary.

File: Branch-SDK/Branch-SDK/Requests/BranchOpenRequest.m (hunk 2 above)
Comment (E-B-Smith) [inline Branch-SDK/Branch-SDK/Requests/BranchOpenRequest.m]: I will clean up the logs with some debug logging.

With the debug logging the block I think the code won't look weird.

**Comment by E-B-Smith** [comment]
@aaustin Wrote: 

> Can you add a test case for when this is called in a non-main thread as well being called from the main thread?

There are test cases in the `synchronous-refer-params-AIS-8-test` branch, including tests for deadlocks.

File: Branch-SDK/Branch-SDK/Branch.m [hunk 8]
```diff
@@ -796,6 +797,20 @@ - (NSDictionary *)getLatestReferringParams {
     return origSessionParams;
 }
 
+- (NSDictionary*) getLatestReferringParamsSynchronous {
+    if (!self.isInitialized &&
+        !self.preferenceHelper.shouldWaitForInit &&
+        ![self.requestQueue containsInstallOrOpen]) {
+		NSLog(@"[Branch] Warning: getLatestReferringParamsSynchronous called before session init.");
+		NSLog(@"[Branch] Make sure Branch `initSessionWithLaunchOptions:`, `handleDeepLink:`, or "
```
Comment (ahmednawar) [inline Branch-SDK/Branch-SDK/Branch.m]: should we return an empty dictionary here better than making the app wait forever in case of user error?

File: Branch-SDK/Branch-SDK/Requests/BranchOpenRequest.m [hunk 9]
```diff
@@ -173,4 +176,52 @@ - (NSString *)getActionName {
     return @"open";
 }
 
+
+#pragma - Open Response Lock Handling
+
+
+//	Instead of semaphores, the lock is handled by scheduled dispatch_queues.
```
Comment (ahmednawar) [inline Branch-SDK/Branch-SDK/Requests/BranchOpenRequest.m]: nitpick: is a tab needed between the `//` and the comment text? on a broader note, perhaps we could start thinking of having a linter for the sdk and decide on a style.

File: Branch-SDK/Branch-SDK/BNCDeviceInfo.m [hunk 10]
```diff
@@ -86,22 +86,22 @@ - (id)init {
     static NSString* browserUserAgentString = nil;
 
     void (^setUpBrowserUserAgent)() = ^() {
-        browserUserAgentString =
-            [[[UIWebView alloc]
-              initWithFrame:CGRectZero]
-                stringByEvaluatingJavaScriptFromString:@"navigator.userAgent"];
+		if (!browserUserAgentString) {
```
Comment (ahmednawar) [inline Branch-SDK/Branch-SDK/BNCDeviceInfo.m]: This is out of scope for this PR, but I would propose checking this value every 30 days similar to what we do with strong matching instead with every request and cache it in the pref helper. AFAIU, this value will not change unless the user updates their OS. Allocating a `WebView` for every single request is a bit expensive.

File: Branch-SDK/Branch-SDK/Requests/BranchOpenRequest.m (hunk 9 above)
Comment (E-B-Smith) [inline Branch-SDK/Branch-SDK/Requests/BranchOpenRequest.m]: By habit I use a tab.  I frequently see people use two spaces after the comment delimiter too.

On a broader note, I started an experimental SDK project that has a clang-format definition file.  The format file is called '.clang-format'.

The project is here:  https://github.com/BranchMetrics/BranchLabs-iOS

We can play with formatting there before deciding.  But I think having a formatter is a good idea.

File: Branch-SDK/Branch-SDK/BNCDeviceInfo.m [hunk 11]
```diff
@@ -83,27 +83,60 @@ - (id)init {
...
 
-    return self;
+    static NSString* browserUserAgentString = nil;
+	void (^setBrowserUserAgent)() = ^() {
+		if (!browserUserAgentString) {
+			browserUserAgentString =
+				[[[UIWebView alloc]
+				  initWithFrame:CGRectZero]
+					stringByEvaluatingJavaScriptFromString:@"navigator.userAgent"];
+			NSLog(@"[Branch] userAgentString: '%@'.", browserUserAgentString);
```
Comment (aaustin) [inline Branch-SDK/Branch-SDK/BNCDeviceInfo.m]: We don't want this log in there

File: Branch-SDK/Branch-SDK/BNCDeviceInfo.m [hunk 12]
```diff
@@ -83,27 +83,60 @@ - (id)init {
...
+		return browserUserAgentString;
+	}
+
+	//	Wait and yield to prevent deadlock:
+
+	dispatch_block_t agentBlock = dispatch_block_create_with_qos_class(
+		DISPATCH_BLOCK_DETACHED | DISPATCH_BLOCK_ENFORCE_QOS_CLASS,
+		QOS_CLASS_USER_INTERACTIVE,
+		0,  ^ {
+			NSLog(@"Will do.");
```
Comment (aaustin) [inline Branch-SDK/Branch-SDK/BNCDeviceInfo.m]: or these logs - this and 2 below

File: Branch-SDK/Branch-SDK/BNCServerInterface.m (hunk 1 above)
Comment (aaustin) [inline Branch-SDK/Branch-SDK/BNCServerInterface.m]: Did you see this comment?

File: Branch-SDK/Branch-SDK/BNCServerInterface.m [hunk 13]
```diff
@@ -138,10 +138,9 @@ - (void)genericHTTPRequest:(NSURLRequest *)request retryNumber:(NSInteger)retryN
             }
             
         }
-        dispatch_async(dispatch_get_main_queue(), ^{
-            if (callback)
-                callback(serverResponse, error);
-        });
+		//	Don't call on the main queue since it might be blocked.
```
Comment (aaustin) [inline Branch-SDK/Branch-SDK/BNCServerInterface.m]: This callback block has traditionally always been handled on the main thread. Changing this could break a lot of existing integrations that might assume we're always calling their code on the main thread.

Comment (aaustin) [inline Branch-SDK/Branch-SDK/BNCServerInterface.m]: Oh - I see you've moved it further down into `initSessionCallback`. Will delete this comment.

File: Branch-SDK/Branch-SDK/Branch.m (hunk 5 above)
Comment (aaustin) [inline Branch-SDK/Branch-SDK/Branch.m]: Please confirm this ^

File: Branch-SDK/Branch-SDK/Requests/BranchOpenRequest.m [hunk 14]
```diff
@@ -173,4 +177,53 @@ - (NSString *)getActionName {
...
+    if (self != [BranchOpenRequest self])
+        return;
+    openRequestWaitQueue =
+        dispatch_queue_create("io.branch.sdk.openqueue", DISPATCH_QUEUE_CONCURRENT);
+}
+
++ (void) setWaitNeededForOpenResponseLock {
+    @synchronized (self) {
+        if (!openRequestWaitQueueIsSuspended) {
+            NSLog(@"Suspend openRequestWaitQueue.");
```
Comment (aaustin) [inline Branch-SDK/Branch-SDK/Requests/BranchOpenRequest.m]: Still a lot of logs in the code

File: Branch-SDK/Branch-SDK/Requests/BranchOpenRequest.m (hunk 6 above)
Comment (aaustin) [inline Branch-SDK/Branch-SDK/Requests/BranchOpenRequest.m]: Won't the call to self.callback(NO, error) on line 88 passthrough that network error? I believe it will but you should confirm while testing with no network & network attenuation

**Comment by aaustin** [comment]
Hey - you can review my outstanding comments by clicking the "Files changed" tab https://github.com/BranchMetrics/ios-branch-deep-linking/pull/536/files. Let's address each one before proceeding.

Lastly, can you update the README with the new API?

File: Branch-SDK/Branch-SDK/BNCDeviceInfo.m (hunk 11 above)
Comment (E-B-Smith) [inline Branch-SDK/Branch-SDK/BNCDeviceInfo.m]: I'll remove it after the final QA round.  It'll be much easier to see if this is breaking with the logging in.

File: Branch-SDK/Branch-SDK/BNCServerInterface.m (hunk 1 above)
Comment (E-B-Smith) [inline Branch-SDK/Branch-SDK/BNCServerInterface.m]: Would you prefer it all on one line?

File: Branch-SDK/Branch-SDK/Requests/BranchOpenRequest.m (hunk 14 above)
Comment (E-B-Smith) [inline Branch-SDK/Branch-SDK/Requests/BranchOpenRequest.m]: Again, I'll remove them after the final QA pass.

File: Branch-SDK/Branch-SDK/Branch.m (hunk 5 above)
Comment (E-B-Smith) [inline Branch-SDK/Branch-SDK/Branch.m]: The call to `setWaitNeededForOpenResponseLock` in initialize sets the lock because there are no `latestReferringParams` to return yet.

Later I set it because there will be some `latestReferringParams` to return, but we haven't received them yet.

Calling `setWaitNeededForOpenResponseLock` twice does no harm.

File: Branch-SDK/Branch-SDK/Requests/BranchOpenRequest.m (hunk 6 above)
Comment (E-B-Smith) [inline Branch-SDK/Branch-SDK/Requests/BranchOpenRequest.m]: I'll test for it.

File: Branch-SDK/Branch-SDK/BNCServerInterface.m (hunk 1 above)
Comment (aaustin) [inline Branch-SDK/Branch-SDK/BNCServerInterface.m]: No it's not that. See line 140? It's got 8 spaces whereas these two lines have 2 tabs. It's just a weird indentation thing for these 2 lines alone.

Comment (E-B-Smith) [inline Branch-SDK/Branch-SDK/BNCServerInterface.m]: Ah.  I must have an editor that is converting spaces to tabs or vice versa.  I can fix that.

**Comment by aaustin** [comment]
Ok I think I'm comfortable for this to move to QA. Please please please please please remember to move the logs. There's nothing worse than an SDK that print debug logs in production apps. 👍
